File: model-client/src/main/java/org/modelix/model/operations/SetPropertyOp.java
package org.modelix.model.operations;

import de.q60.mps.shadowmodels.runtime.model.persistent.IWriteTransaction;
import java.text.MessageFormat;

public class SetPropertyOp extends AbstractOperation {

  public final long nodeId;
  public final String role;
  public final String value;

  public SetPropertyOp(long nodeId, String role, String value) {
    this.nodeId = nodeId;
    this.role = role;
    this.value = value;
  }

  @Override
  public IAppliedOperation apply(IWriteTransaction transaction) {
    String oldValue = transaction.getProperty(nodeId, role);
    transaction.setProperty(nodeId, role, value);
    return new Applied(oldValue);
  }

  @Override
  public IOperation transform(IOperation previous) {
    if (previous instanceof SetPropertyOp) {
      return this;
    } else if (previous instanceof SetReferenceOp) {
      return this;
    } else if (previous instanceof AddNewChildOp) {
      return this;
    } else if (previous instanceof DeleteNodeOp) {
      DeleteNodeOp o = ((DeleteNodeOp) previous);
      if (nodeId == o.childId) {
        return new NoOp();
      } else {
        return this;
      }
    } else if (previous instanceof MoveNodeOp) {
      return this;
    } else {
      throw new RuntimeException("Unknown type: " + previous.getClass().getName());
    }
  }

  @Override
  public String toString() {
    return MessageFormat.format("SetPropertOp {0}.{1} = {2}", Long.toHexString(nodeId), role, value);
  }

  public class Applied extends AbstractOperation.Applied implements IAppliedOperation {
    private String oldValue;

    public Applied(String oldValue) {
      this.oldValue = oldValue;
    }

    @Override
    public IOperation getOriginalOp() {
      return SetPropertyOp.this;
    }

    @Override
    public IOperation invert() {
      return new SetPropertyOp(nodeId, role, oldValue);
    }

    @Override
    public String toString() {
      return super.toString() + ", oldValue: " + oldValue;
    }
  }
}

File: model-client/src/main/kotlin/org/modelix/model/operations/SetPropertyOp.kt
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License. 
 */

package org.modelix.model.operations

import org.modelix.model.api.IWriteTransaction
import java.text.MessageFormat

class SetPropertyOp(val nodeId: Long, val role: String, val value: String) : AbstractOperation() {
    override fun apply(transaction: IWriteTransaction?): IAppliedOperation? {
        val oldValue = transaction!!.getProperty(nodeId, role)
        transaction.setProperty(nodeId, role, value)
        return Applied(oldValue!!)
    }

    override fun transform(previous: IOperation?): IOperation? {
        return if (previous is SetPropertyOp) {
            this
        } else if (previous is SetReferenceOp) {
            this
        } else if (previous is AddNewChildOp) {
            this
        } else if (previous is DeleteNodeOp) {
            if (nodeId == previous.childId) {
                NoOp()
            } else {
                this
            }
        } else if (previous is MoveNodeOp) {
            this
        } else {
            throw RuntimeException("Unknown type: " + previous!!.javaClass.getName())
        }
    }

    override fun toString(): String {
        return MessageFormat.format("SetPropertOp {0}.{1} = {2}", java.lang.Long.toHexString(nodeId), role, value)
    }

    inner class Applied(private val oldValue: String) : AbstractOperation.Applied(), IAppliedOperation {
        override val originalOp: IOperation
            get() = this@SetPropertyOp

        override fun invert(): IOperation {
            return SetPropertyOp(nodeId, role, oldValue)
        }

        override fun toString(): String {
            return super.toString() + ", oldValue: " + oldValue
        }
    }
}
